File: benchmarks/06_logparse.py
import sys

info = warn = error = debug = 0
auth = api = db = cache = scheduler = 0
total = 0

with open("benchmarks/sample_large.log", "r") as f:
    for line in f:
        total += 1
        if "[INFO]" in line: info += 1
        if "[WARN]" in line: warn += 1
        if "[ERROR]" in line: error += 1
        if "[DEBUG]" in line: debug += 1
        if "(auth)" in line: auth += 1
        if "(api)" in line: api += 1
        if "(db)" in line: db += 1
        if "(cache)" in line: cache += 1
        if "(scheduler)" in line: scheduler += 1

print(f"total lines: {total}")
print(f"INFO: {info}, WARN: {warn}, ERROR: {error}, DEBUG: {debug}")
print(f"auth: {auth}, api: {api}, db: {db}, cache: {cache}, scheduler: {scheduler}")

File: benchmarks/06_logparse.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int main() {
    FILE *f = fopen("benchmarks/sample_large.log", "r");
    if (!f) {
        fprintf(stderr, "cannot open log file\n");
        return 1;
    }

    int info = 0, warn = 0, error = 0, debug = 0;
    int auth = 0, api = 0, db = 0, cache = 0, scheduler = 0;
    int total = 0;

    char buf[4096];
    while (fgets(buf, sizeof(buf), f)) {
        total++;
        if (strstr(buf, "[INFO]")) info++;
        if (strstr(buf, "[WARN]")) warn++;
        if (strstr(buf, "[ERROR]")) error++;
        if (strstr(buf, "[DEBUG]")) debug++;
        if (strstr(buf, "(auth)")) auth++;
        if (strstr(buf, "(api)")) api++;
        if (strstr(buf, "(db)")) db++;
        if (strstr(buf, "(cache)")) cache++;
        if (strstr(buf, "(scheduler)")) scheduler++;
    }
    fclose(f);

    printf("total lines: %d\n", total);
    printf("INFO: %d, WARN: %d, ERROR: %d, DEBUG: %d\n", info, warn, error, debug);
    printf("auth: %d, api: %d, db: %d, cache: %d, scheduler: %d\n", auth, api, db, cache, scheduler);
    return 0;
}
